README.md
#About#

This is a collection of compression tools, and a script to download/build them, using the Android NDK.

- gzip 1.6 - GPLv3
- bzip2 1.0.6 - BSD-ish
- xz utils (git) - public domain core / (L)GPLv2/v3 portions
- lz4 (git) - BSD 2-clause core / GPLv2  portions
- lzo 2.09 - GPLv2
- lzop 1.03 - GPLv2

I make no claims as to how well the output programs in all their variants actually work. I only made sure they built. I need this for a different project, but thought I might share it. If anything breaks, you get to keep all the pieces.

#Building#

At least the following packages are needed:

- automake
- autoconf
- autopoint
- autotools-dev
- m4
- perl

Versions needed may vary.

I built all of this on an x86-64 Linux Mint box.

To download fresh copies, run ``./download.sh``. Note that this script makes some minor modifications to the downloads. Because upstream may change, I included all files as they were at the time of this writing.

To build, run ``./build.sh``, and wait approximately one forever.



gzip/NEWS
GNU gzip NEWS                                    -*- outline -*-

* Noteworthy changes in release 1.6 (2013-06-09) [stable]

** New features

  gzip now accepts the --keep (-k) option, for consistency with tools
  like xz, lzip and bzip2.  With this option, gzip no longer removes
  named input files when compressing or decompressing.

** Bug fixes

  gzip -d no longer malfunctions with certain invalid data in 'pack' format.
  [bug introduced in gzip-0.8]

  When overwriting, gzip no longer acts as if you typed "y" when you type "n",
  on some platforms when compiled with optimization.
  [bug introduced in gzip-1.3.6]

  zgrep no longer malfunctions with a multi-digit context option like -15.
  Now, it passes that option to grep (equivalent to -C15) just as it does
  for single-digit options. [bug introduced in gzip-1.3.12]

  zmore now acts more like 'more', and is more portable to POSIXish hosts.


* Noteworthy changes in release 1.5 (2012-06-17) [stable]

** Bug fixes

  gzip -d now decodes and checks header CRC16 checksums as specified by
  the FHCRC section of Internet RFC 1952.

  "gzip -d -S '' precious.gz" is now rejected immediately.  Before,
  that command would emulate "rm -i precious.gz", but with an easily-
  misunderstood prompt.  I.e., gzip would ask if it's ok to remove the
  existing file, "precious.gz".  If you made the mistake of saying "yes",
  it would remove that input file before attempting to uncompress it.

  gzip -cdf now properly handles input consisting of gzip'd data followed
  by uncompressed data.  Before it would output raw compressed input, too.
  For example, now "(printf x|gzip; echo y)|gzip -dcf" prints "xy\n",
  while before it would print "x<compressed data>y\n".

  gzip -rf no longer compresses files more than once (e.g., replacing
  FOO with FOO.gz.gz) on file systems such as ZFS where a readdir
  loop that unlinks and creates files can revisit output files.


* Noteworthy changes in release 1.4 (2010-01-20) [stable]

** Bug fixes

  gzip -d could segfault and/or clobber the stack, possibly leading to
  arbitrary code execution.  This affects x86_64 but not 32-bit systems.
  This fixes CVE-2010-0001.
  For more details, see http://bugzilla.redhat.com/554418

  gzip -d would fail with a CRC error for some valid inputs.
  So far, the only valid input known to exhibit this failure was
  compressed "from FAT filesystem (MS-DOS, OS/2, NT)".  In addition,
  to trigger the failure, your memcpy implementation must copy in
  the "reverse" order.


* Noteworthy changes in release 1.3.14 (2009-10-30) [beta]

** Bug fixes

  gzip no longer fails when there is exactly one trailing NUL byte
  gzip has always accepted trailing NUL bytes.  Note the plural.

  zdiff would exit with status 2 (indicating an error) rather than 1 to
  indicate differences when both inputs were compressed and different.

  zdiff would fail to print differences in two compressed inputs

  zgrep -f - didn't work


* Noteworthy changes in release 1.3.13 (2009-09-30) [stable]

** 'gzip -f foo.gz' now creates a file foo.gz.gz instead of complaining.

** Bug fixes

  gzip -d no longer fails with "-" as 2nd or subsequent argument

  gzip no longer ignores a close-induced write failure, e.g., on NFS

  gzip -d no longer segfaults on certain invalid inputs


Major changes in Gzip 1.3.12 (2007-04-13)

* znew now uses $TMPDIR (default /tmp) instead of always using /tmp.

* It is now documented that gzip ignores case when examining file name
  extensions; for example, 'gzip test.Gz' (without -f) fails because
  the file name ends in '.Gz'.

Major changes in Gzip 1.3.11 (2007-02-05)

* As per the GNU coding standards, the behavior of gzip and its
  companion executables no longer depend on the name used to invoke them.
  For example, 'gzip' and 'gunzip' are no longer hard links;
  instead, 'gunzip' is now a small program that invokes 'gzip -d'.

* zdiff now checks for subsidiary gzip failures, and works around
  bugs in IRIX 6 sh, Tru64 4.0F ksh, and Solaris 8 bash.

Major changes in Gzip 1.3.10 (2006-12-30)

* gzip -c and zcat now work on special files, files with special mode bits,
  and files with multiple hard links.
* gzip -q now exits with status 2 (not 1) when SIGPIPE is received.
* zcmp and zdiff did not work in the usual case, due to a typo.
* zgrep has many bugs fixed with argument handling, special characters,
  and exit status.
* zless no longer mishandles $%=~ in file names.

Gzip 1.3.9 (2006-12-15)

* No major changes; only porting fixes.

Major changes in Gzip 1.3.8 (2006-12-08)

* Fix some gzip problems:
  - A security fix from Debian 1.3.5-5 was inadvertently omitted.
  - The assembler is now invoked with --noexecstack if supported,
    so that gzip can better resist stack-smashing attacks.

Major changes in Gzip 1.3.7 (2006-12-06)

* Fix some gzip problems:
  - Refuse to compress setuid or setgid files, or files with the sticky bit.
  - Fix more race conditions in setting file permissions and owner,
    removing output files, following symbolic links, and dealing with
    special files.
  - Remove most of the code working around ENAMETOOLONG deficiencies.
    Systems with those deficiencies are long-dead, and the workarounds
    had race conditions on modern hosts.
  - Catch CPU time and file size limit signals, too.
  - Check for read errors when closing files.
  - Fix a core dump caused by a stray abort mistakenly introduced in 1.3.6.
* Fix some gzexe problems:
  - Improve resistance to denial-of-service attacks.
  - Fix some quoting and escaping bugs.
  - Do not assume /tmp is sticky (though it should be!).
  - Do not assume the working directory can be written.
  - Rely on PATH in the generated executable, as the man page says.
  - Don't assume IFS is sane.
  - Exit with signal's status, if signaled.

Major changes in Gzip 1.3.6 (2006-11-20)

* Fix some race conditions in setting file time stamps, permissions, and owner.
* Fix some race conditions in signal handling.

download.sh
# gzip

if [ -d "gzip" ]; then rm -rf gzip; fi
wget http://ftp.gnu.org/gnu/gzip/gzip-1.6.tar.gz
tar xvf gzip-1.6.tar.gz
mv gzip-1.6 gzip
rm gzip-1.6.tar.gz

# gnulib

if [ -d "gnulib" ]; then rm -rf gnulib; fi
git clone git://git.savannah.gnu.org/gnulib.git

# patch gnulib in gzip with newer version that supports Android, yikes

if [ -d "gnulib/m4" ]; then
  cd gnulib/m4
  for i in `ls`; do
    if [ -f "../../gzip/m4/$i" ]; then
      cp -f $i ../../gzip/m4/$i
    fi
  done
  cd ../..
fi

if [ -d "gnulib" ]; then
  if [ -d "gzip" ]; then
    cp -f gnulib/build-aux/ar-lib gzip/build-aux/ar-lib
    cp -f gnulib/m4/absolute-header.m4 gzip/m4/absolute-header.m4
    cp -f gnulib/lib/stdio-impl.h gzip/lib/stdio-impl.h
    cd gzip
    autoreconf -fi
    cd lib
    cat Makefile.in | grep "fpending.\$(OBJEXT)" >/dev/null
    if [ "$?" -ne "0" ]; then
      sed -i 's/yesno.\$(OBJEXT)/yesno.\$(OBJEXT) fpending.\$(OBJEXT)/g' Makefile.in
    fi
    cd ../..
  fi
fi

# bzip2

if [ -d "bzip2" ]; then rm -rf bzip2; fi
wget http://www.bzip.org/1.0.6/bzip2-1.0.6.tar.gz
tar xvf bzip2-1.0.6.tar.gz
mv bzip2-1.0.6 bzip2
rm bzip2-1.0.6.tar.gz

# crosscompile variables are hardcoded in the bzip2 makefile, breaking the builds

if [ -f "bzip2/Makefile" ]; then
  sed -i '/CC=/d' bzip2/Makefile
  sed -i '/AR=/d' bzip2/Makefile
  sed -i '/RANLIB=/d' bzip2/Makefile
fi

# xz

if [ -d "xz" ]; then rm -rf xz; fi
git clone http://git.tukaani.org/xz.git xz
rm -rf xz/.git

# lz4

if [ -d "lz4" ]; then rm -rf lz4; fi
git clone https://github.com/Cyan4973/lz4.git lz4
rm -rf lz4/.git

# lzo library

if [ -d "lzo" ]; then rm -rf lzo; fi
wget http://www.oberhumer.com/opensource/lzo/download/lzo-2.09.tar.gz
tar xvf lzo-2.09.tar.gz
mv lzo-2.09 lzo
rm lzo-2.09.tar.gz

# lzop (uses lzo)

if [ -d "lzop" ]; then rm -rf lzop; fi
wget http://www.lzop.org/download/lzop-1.03.tar.gz
tar xvf lzop-1.03.tar.gz
mv lzop-1.03 lzop
rm lzop-1.03.tar.gz

# fix lzop compilation for arm64, configure script does not support aarch64, and the tools used to build it are older than latest, so autoreconf barfs

if [ -f "lzop/autoconf/config.sub" ]; then
  cat lzop/autoconf/config.sub | grep "aarch64" >/dev/null
  if [ "$?" -ne "0" ]; then
    sed -i '/a29k \\/a \ \ \ \ \ \ \ \ | aarch64 | aarch64_be \\' lzop/autoconf/config.sub
    sed -i '/a29k-* \\/a \ \ \ \ \ \ \ \ | aarch64-* | aarch64_be-* \\' lzop/autoconf/config.sub
  fi
fi

# clear git

rm -rf */.git
